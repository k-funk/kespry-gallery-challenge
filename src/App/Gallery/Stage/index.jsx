import React from 'react';
import { PropTypes as T } from 'prop-types';

import './style.css';


const Stage = ({ img, selectNextImage, selectPreviousImage }) => {
  if (!img) { return null; }

  return (
    <div className="stage">
      <div
        className="select-image previous"
        role="button"
        onClick={selectPreviousImage}
        onKeyUp={selectPreviousImage}
        tabIndex={0}
      >
        ❮
      </div>
      <div
        className="select-image next"
        role="button"
        onClick={selectNextImage}
        onKeyUp={selectNextImage}
        tabIndex={0}
      >
        ❯
      </div>
      <img
        src={img}
        alt={img}
      />
    </div>
  );
};

Stage.propTypes = {
  img: T.string,
  selectNextImage: T.func.isRequired,
  selectPreviousImage: T.func.isRequired,
};

export default Stage;
