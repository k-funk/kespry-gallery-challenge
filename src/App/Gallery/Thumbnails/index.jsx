import React from 'react';
import { PropTypes as T } from 'prop-types';

import './style.css';


const Thumbnails = ({ imageList, selectImage }) => (
  <div className="thumbnails">
    {imageList.map((imageUrl, idx) => (
      <span
        key={imageUrl}
        role="button"
        tabIndex={0}
        onClick={() => selectImage(imageUrl)}
        onKeyUp={() => selectImage(imageUrl)}
      >
        <img
          src={imageUrl}
          alt={`Thumbnail ${idx}`}
          className="thumbnail"
        />
      </span>
    ))}
  </div>
);

Thumbnails.propTypes = {
  imageList: T.arrayOf(T.string).isRequired,
  selectImage: T.func.isRequired,
};

export default Thumbnails;
