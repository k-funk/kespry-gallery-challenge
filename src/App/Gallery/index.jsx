import React, { useState } from 'react';
import { PropTypes as T } from 'prop-types';
import Thumbnails from './Thumbnails';
import Stage from './Stage';


const Gallery = ({ imageList }) => {
  const [stageImg, setStageImg] = useState('');

  const selectNextImage = () => {
    const currentIdx = imageList.indexOf(stageImg);
    const nextIdx = currentIdx >= imageList.length - 1 ? 0 : currentIdx + 1;
    return setStageImg(imageList[nextIdx]);
  };

  const selectPreviousImage = () => {
    const currentIdx = imageList.indexOf(stageImg);
    const previousIdx = currentIdx === 0 ? imageList.length - 1 : currentIdx - 1;
    return setStageImg(imageList[previousIdx]);
  };

  return (
    <div className="card">
      {!imageList.length ? (
        <div>No images in gallery.</div>
      ) : (
        <>
          <Thumbnails imageList={imageList} selectImage={setStageImg} />
          <Stage
            img={stageImg}
            selectNextImage={selectNextImage}
            selectPreviousImage={selectPreviousImage}
          />
        </>
      )}
    </div>
  );
};

Gallery.propTypes = {
  imageList: T.arrayOf(T.string).isRequired,
};

export default Gallery;
