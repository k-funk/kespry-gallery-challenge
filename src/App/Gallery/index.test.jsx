import React from 'react';
import { shallow } from 'enzyme';
import shallowToJson from 'enzyme-to-json';

import Gallery from './index';


describe('outputs the expected tree when', () => {
  let wrapper;
  let setStageImgSpy;
  let defaultProps;

  beforeEach(() => {
    setStageImgSpy = jest.fn();
    defaultProps = {
      stageImg: 'https://img.com/1',
      setStageImg: setStageImgSpy,
    };
  });

  test('images in gallery', () => {
    wrapper = shallow((
      <Gallery
        {...defaultProps}
        imageList={['https://img.com/1', 'https://img.com/2']}
      />
    ));
  });

  test('no images in gallery', () => {
    wrapper = shallow((
      <Gallery
        {...defaultProps}
        imageList={[]}
      />
    ));
  });

  afterEach(() => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
