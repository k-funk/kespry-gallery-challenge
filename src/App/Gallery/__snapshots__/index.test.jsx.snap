// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outputs the expected tree when images in gallery 1`] = `
<div
  className="card"
>
  <Thumbnails
    imageList={
      Array [
        "https://img.com/1",
        "https://img.com/2",
      ]
    }
    selectImage={[MockFunction]}
  />
  <Stage
    img="https://img.com/1"
    selectNextImage={[Function]}
    selectPreviousImage={[Function]}
  />
</div>
`;

exports[`outputs the expected tree when no images in gallery 1`] = `
<div
  className="card"
>
  <div>
    No images in gallery.
  </div>
</div>
`;
